src/app/dashboard/constraint-builder-overlay/constraint-builder.component.ts

```diff
@@ -0,0 +1,163 @@
+
+  constraint?: ConstraintWrapper;
+
+  private students: Student[];
+  private skills: Skill[];
+  private projects: Project[];
+  private subscriptions: Subscription[] = [];
+
+  private constraintFunctions: ConstraintFunction[];
+  private readonly CONSTRAINT_OPERATORS = [Operator.GREATER_THAN_OR_EQUAL, Operator.LESS_THAN_OR_EQUAL];
+
+  constructor(
+    private constraintsService: ConstraintsService,
+    private overlayService: OverlayService,
+    private constraintBuilderService: ConstraintBuilderService,
+    private studentsService: StudentsService,
+    private skillsService: SkillsService,
+    private projectsService: ProjectsService
+  ) {}
+
+  ngOnInit(): void {
+    this.subscriptions.push(
+      this.studentsService.students$.subscribe(students => {
+        this.students = students;
+        this.updateData();
+      }),
+      this.skillsService.skills$.subscribe(skills => {
+        this.skills = skills;
+        this.updateData();
+      }),
+      this.projectsService.projects$.subscribe(projects => {
+        this.projects = projects;
+        this.updateData();
+      })
+    );
+  }
+
+  ngOnDestroy(): void {
+    this.subscriptions.forEach(subscription => subscription.unsubscribe());
+  }
+
+  private updateData(): void {
+    if (!this.students || !this.skills || !this.projects) return;
+    this.constraintFunctions = this.constraintBuilderService.getConstraintFunctions(this.students, this.skills);
+    console.log(this.constraintFunctions);
+    this.constraintFunctionProperties = this.constraintBuilderService.getConstraintFunctionProperties(
+      this.constraintFunctions
+    );
+    this.constraintProjectSelectData = this.constraintBuilderService.getConstraintProjects(this.projects);
+    this.selectedConstraintProject = this.constraintProjectSelectData[0].id;
+    this.constraintOperators = this.constraintBuilderService.getConstraintOperators();
+    this.selectedConstraintOperator = this.constraintOperators[0].id;
+  }
+
+  selectData(): void {
+    this.getConstraintFunctionValues();
+    this.getConstraintFunctionOperators();
+    this.getConstraint();
+  }
+
+  private getConstraintFunctionValues(): void {
+    this.constraintFunctionValueSelectData = this.constraintBuilderService.getConstraintFunctionValues(
+      this.constraintFunctions,
+      this.selectedConstraintFunctionProperty
+    );
+
+    if (!this.constraintFunctionValueSelectData || !this.constraintFunctionValueSelectData.length) {
+      this.selectedConstraintFunctionValue = null;
+    }
+    this.selectedConstraintFunctionValue = this.constraintFunctionValueSelectData[0].id;
+  }
+
+  private getConstraintFunctionOperators(): void {
+    this.constraintFunctionOperatorSelectData = this.constraintBuilderService.getConstraintFunctionOperators(
+      this.constraintFunctions,
+      this.selectedConstraintFunctionProperty
+    );
+
+    if (!this.constraintFunctionOperatorSelectData || !this.constraintFunctionOperatorSelectData.length) {
```

Comment:
` this.selectedConstraintFunctionOperator = !this.constraintFunctionOperatorSelectData?.length ? this.constraintFunctionOperatorSelectData[0].id : null`